src/tensor/norm_fiber.cc

```diff
@@ -0,0 +1,182 @@
+/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
+ *                              (Skoltech), Russia. All rights reserved.
+ *                 2023-present Artificial Intelligence Research Institute
+ *                              (AIRI), Russia. All rights reserved.
+ *
+ * NNTile is software framework for fast training of big neural networks on
+ * distributed-memory heterogeneous systems based on StarPU runtime system.
+ *
+ * @file src/tensor/norm_fiber.cc
+ * Euclidean norms over slices into a fiber of a product of a Tensor<T>
+ *
+ * @version 1.0.0
+ * */
+
+#include "nntile/tensor/norm_fiber.hh"
+#include "nntile/starpu/norm_fiber.hh"
+
+namespace nntile::tensor
+{
+
+//! Tensor-wise norm_fiber
+template<typename T>
+void norm_fiber_async(Scalar alpha, const Tensor<T> &src, Scalar beta,
+        const Tensor<T> &dst, Index axis, Index batch_ndim, int redux)
+{
+    // Check dimensions
+    if(dst.ndim != batch_ndim+1)
+    {
+        throw std::runtime_error("dst.ndim != batch_ndim+1");
+    }
+    // Treat special case of src.ndim=0
+    if(src.ndim == 0)
+    {
+        throw std::runtime_error("Scalar input makes no sense");
+    }
+    // Check axis
+    if(axis < 0)
+    {
+        throw std::runtime_error("axis < 0");
+    }
+    if(axis >= src.ndim-batch_ndim)
+    {
+        throw std::runtime_error("axis >= src.ndim-batch_ndim");
+    }
+    // Check shapes
+    if(dst.shape[0] != src.shape[axis])
+    {
+        throw std::runtime_error("dst.shape[0] != src.shape[axis]");
+    }
+    if(dst.basetile_shape[0] != src.basetile_shape[axis])
+    {
+        throw std::runtime_error("dst.basetile_shape[0] != "
+                "src.basetile_shape[axis]");
+    }
+    for(Index i = 0; i < batch_ndim; ++i)
```

Comment:
This cycle is incorrect. If `batch_ndim=0`, then this check does nothing. There must be 2 separate check cycles, take a look at `sum_fiber` function.